runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -89,6 +89,151 @@ typedef struct iree_hal_hip_device_t {
+  hipEvent_t hip_event = (hipEvent_t)(event);
+  return IREE_HIP_RESULT_TO_STATUS(
+      table->hip_symbols, hipEventRecord(hip_event, table->dispatch_hip_stream),
+      "hipEventRecord");
+}
+
+iree_status_t
+iree_hal_hip_deferred_work_queue_symbol_table_synchronize_native_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    void* event) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  hipEvent_t hip_event = (hipEvent_t)(event);
+  return IREE_HIP_RESULT_TO_STATUS(table->hip_symbols,
+                                   hipEventSynchronize(hip_event));
+}
+iree_status_t
+iree_hal_hip_deferred_work_queue_symbol_table_destroy_native_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    void* event) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  hipEvent_t hip_event = (hipEvent_t)(event);
+  return IREE_HIP_RESULT_TO_STATUS(table->hip_symbols,
+                                   hipEventDestroy(hip_event));
+}
+
+iree_status_t
+iree_hal_hip_deferred_work_queue_symbol_table_semaphore_acquire_timepoint_device_signal_native_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    struct iree_hal_semaphore_t* semaphore, uint64_t value, void** out_event) {
+  hipEvent_t* out = (hipEvent_t*)(out_event);
+  return iree_hal_hip_event_semaphore_acquire_timepoint_device_signal(
+      semaphore, value, out);
+}
+
+bool iree_hal_hip_deferred_work_queue_symbol_table_acquire_host_wait_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    struct iree_hal_semaphore_t* semaphore, uint64_t value, void** out_event) {
+  return iree_hal_hip_semaphore_acquire_event_host_wait(
+      semaphore, value, (iree_hal_hip_event_t**)out_event);
+}
+
+void iree_hal_hip_deferred_work_queue_symbol_table_release_wait_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    void* wait_event) {
+  iree_hal_hip_event_release(wait_event);
+}
+
+void* iree_hal_hip_deferred_work_queue_symbol_table_native_event_from_wait_event(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    void* event) {
+  iree_hal_hip_event_t* wait_event = (iree_hal_hip_event_t*)event;
+  return iree_hal_hip_event_handle(wait_event);
+}
+
+iree_status_t
+iree_hal_hip_deferred_work_queue_symbol_table_create_command_buffer_for_deferred(
+    iree_hal_deferred_work_queue_device_interface_t* symbol_table,
+    iree_hal_command_buffer_mode_t mode, iree_hal_command_category_t categories,
+    iree_hal_command_buffer_t** out) {
+  iree_hal_hip_deferred_work_queue_device_interface_t* table =
+      (iree_hal_hip_deferred_work_queue_device_interface_t*)(symbol_table);
+  return iree_hal_hip_device_create_stream_command_buffer(table->device, mode,
```

Comment:
it was always a bug that there was a new stream command buffer created - we should be reusing the same one ideally - in a follow up we should try to remove this

Reply:
So now that I have hung some additional data off the command buffers (related to picking up tracing data) we can re-use but only when they are not concurrent uses, so we might need a small pool.

Reply:
I think moving your state to an independent struct that you pass in to the stream command buffer would let you decouple things (instead of keeping the command buffer around to keep your tracing state, keep your tracing state on the submission and pass it into the command buffer when executing) - this will all change when we get properly reusable command buffers, though, so it's worth waiting until then (at that point the stream command buffer gets deleted)